crates/ruff_python_semantic/src/model.rs

```diff
@@ -1575,6 +1575,31 @@ impl<'a> SemanticModel<'a> {
             .intersects(SemanticModelFlags::DEFERRED_TYPE_DEFINITION)
     }
 
+    pub const fn in_valid_context_for_deferred_type_definition(&self) -> bool {
+        if self.in_typing_literal() {
+            return false;
+        }
+        if !self.in_type_definition() {
+            return false;
+        }
+        // `in_deferred_type_definition()` will only be `true` if we're now visiting the deferred nodes
+        // after having already traversed the source tree once. If we're now visiting the deferred nodes,
+        // we can't defer again, or we'll infinitely recurse!
+        if self.in_deferred_type_definition() {
+            return false;
+        }
+        if self.in_type_checking_block() || self.flags.intersects(SemanticModelFlags::STUB_FILE) {
```

Comment:
Why not use `in_typing_only_block` that you added in this PR? Currently it's only used in a debug assertion, but it seems like it's equivalent to this line?

Reply:
oops, I think I just added the `in_typing_only_block()` after I added this bit